### PR TITLE
semver: reject leading zeroes and empty pre-release/build

### PR DESCRIPTION
**Pull request by sueun-dev**
### What

`semver.Parse` (behind the `semver.is_valid` and `semver.compare` built-ins) accepts several strings that are not valid SemVer 2.0.0 versions:

```rego
semver.is_valid("01.2.3")         # true, should be false
semver.is_valid("1.02.3")         # true
semver.is_valid("1.2.3-01")       # true  (leading zero in a numeric pre-release id)
semver.is_valid("1.2.3-")         # true  (empty pre-release)
semver.is_valid("1.2.3+")         # true  (empty build metadata)
semver.compare("1.02.3", "1.2.3") # 0, i.e. treated as equal
```

The major/minor/patch numbers are parsed straight through `strconv.ParseInt`, which happily accepts a leading zero, and the pre-release/metadata identifier check is skipped whenever the section is empty (`if v.PreRelease != ""`). The [spec](https://semver.org/#spec-item-2) forbids leading zeroes in the numeric identifiers and requires a non-empty pre-release/build section after the `-`/`+`.

`semver.is_valid` is documented as "Validates that the input is a valid SemVer string", so returning `true` here is wrong, and `semver.compare` silently treats a malformed version as an ordinary one instead of erroring the way it already does for e.g. `1.2`.

### Fix

- Validate major/minor/patch against `0|[1-9][0-9]*` before `ParseInt`.
- Run the identifier check whenever the `-`/`+` separator is present (an empty section is now rejected), using `strings.Cut` for the found flag.
- Reject a leading zero in a numeric pre-release identifier. Build-metadata identifiers keep allowing leading zeroes, which the spec permits.

`reMetaIdentifier` already rejects other spec violations (e.g. `_`), so this brings the numeric-identifier handling in line with that.

### Test

Added the new invalid strings to `TestBadInput` (they parsed cleanly before, error now) and a `TestGoodInput` covering the spec-valid edge cases that must keep parsing — a single-zero pre-release (`1.2.3-0`), an alphanumeric id starting with zero (`1.2.3-0a`), and build metadata with a leading zero (`1.2.3+01`). `go test ./internal/semver/` and `./v1/topdown/` pass.

### CODE REVIEW

File: internal/semver/semver.go [hunk 1]
```diff
@@ -45,13 +50,15 @@ type Version struct {
 func Parse(version string) (v Version, err error) {
 	version = strings.TrimPrefix(version, "v")
 
-	version, v.Metadata = cut(version, '+')
-	if v.Metadata != "" && !reMetaIdentifier.MatchString(v.Metadata) {
+	var found bool
```
Comment (charlieegan3) [inline internal/semver/semver.go]:
```suggestion
	var foundMetadata bool
```

File: internal/semver/semver.go [hunk 2]
```diff
@@ -60,22 +67,53 @@ func Parse(version string) (v Version, err error) {
...
+		return v, fmt.Errorf("invalid patch version: %w", err)
 	}
 
 	return v, nil
 }
 
+// parseNumeric parses a major, minor or patch identifier, rejecting the empty
+// string, a sign or a leading zero (all forbidden by SemVer 2.0.0) before
+// converting to int64.
+func parseNumeric(s string) (int64, error) {
```
Comment (charlieegan3) [inline internal/semver/semver.go:86]: Perhaps not using a regex might be faster here

```

  func parseNumeric(s string) (int64, error) {
        if s == "" || s[0] == '+' || s[0] == '-' || (len(s) > 1 && s[0] == '0') {
                return 0, fmt.Errorf("%q is not a valid numeric identifier", s)
        }
        return strconv.ParseInt(s, 10, 64)
  }


```

**Review by charlieegan3 — COMMENTED** [review]
Hey, thanks for this, two small comments. Some of the checks are failing too, please take a look when you have a moment.

File: internal/semver/semver.go [hunk 3]
```diff
@@ -60,18 +64,51 @@ func Parse(version string) (v Version, err error) {
...
+	for id := range strings.SplitSeq(pre, ".") {
+		if len(id) > 1 && id[0] == '0' && isAllDigits(id) {
+			return false
+		}
+	}
+	return true
+}
 
-	return v, err
+func isAllDigits(s string) bool {
```
Comment (sspaink) [inline internal/semver/semver.go:105]: This is a duplicate of `isAllDecimals` that already lives in this file

**Review by sspaink — COMMENTED** [review]
Could you also add test cases under `v1/test/cases/testdata/{v0,v1}/semver*` testing this new behavior? This will help make sure the Swift and Java implementation behave the same. Thanks!

**Comment by sspaink** [comment]
@sueun-dev thank you for working on this! Seeing as the only blocker was a duplicated function and compliance tests I went ahead and did those changes on your behalf to make sure this change lands.

**Review by sspaink — APPROVED** [review]
